### ISSUE TITLE
[Question]: Why is the platform not determined via the `defaultTargetPlatform` variable?

### ISSUE THREAD

**Issue by ju851zel** [fluttercommunity/plus_plugins#1319]
### What is your question?

I'm developing a flutter app.
I'm mocking the android device platform channel for a unit test.
When running the unit test on a linux machine I cannot override the platform that is used and therefore always the linux platform is selected. Because this method is not calling a method channel I cannot mock it.

My question is, why is the platform determined via the `Platform.isAndroid` etc. methods and not via the `defaultTargetPlatform` in the foundation package?

And would a PR changing that be merges. Then I would create it.

### Checklist before submitting a question

- [X] I Google'd a solution and I couldn't find it
- [X] I searched on StackOverflow for a solution and I couldn't find it
- [X] I read the README.md file of the plugin
- [X] I am using the latest version of the plugin
- [X] All dependencies are up to date with `flutter pub upgrade`
- [X] I did a `flutter clean`
- [X] I tried running the example project

**Comment by miquelbeltran** [comment]
The plugins were originally developed by the Flutter team @ Google, and are now maintained by the community, so we can't answer on certain decisions.

Which plugin do you refer exactly, and can you link the lines of code you are specifically asking for?

**Comment by ju851zel** [comment]
Alright, in this case I would propose to change the implementation of the `deviceInfo` getter in the class `DeviceInfoPlugin`, defined in the file `device_info_plus.dart`
Currently the implementation looks like that
```dart
Future<BaseDeviceInfo> get deviceInfo async {
    if (kIsWeb) {
      return webBrowserInfo;
    } else {
      if (Platform.isAndroid) {
        return androidInfo;
      } else if (Platform.isIOS) {
        return iosInfo;
      } else if (Platform.isLinux) {
        return linuxInfo;
      } else if (Platform.isMacOS) {
        return macOsInfo;
      } else if (Platform.isWindows) {
        return windowsInfo;
      }
    }

    throw UnsupportedError('Unsupported platform');
  }
```

As described above the use of `Platform.isXXX` cannot be mocked easily and to my knowledge was the best way to determine the current plattform before the [`defaultTargetPlatform`](https://api.flutter.dev/flutter/foundation/defaultTargetPlatform.html) property was available.

With the availability of `defaultTargetPlatform` the code above would be changed to something like that.
```dart
Future<BaseDeviceInfo> get deviceInfo async {
    if (kIsWeb) {
      return webBrowserInfo;
    } else {
      if (defaultTargetPlatform == TargetPlatform.android) {
        return androidInfo;
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        return iosInfo;
      } else if (defaultTargetPlatform == TargetPlatform.linux) {
        return linuxInfo;
      } else if (defaultTargetPlatform == TargetPlatform.macOS) {
        return macOsInfo;
      } else if (defaultTargetPlatform == TargetPlatform.windows) {
        return windowsInfo;
      }
    }

    throw UnsupportedError('Unsupported platform');
  }
```

This would create the possibility to set the [`debugDefaultTargetPlatformOverride`](https://api.flutter.dev/flutter/foundation/debugDefaultTargetPlatformOverride.html) variable in testing environments and therefore allow tests to simulate different environments.

**Comment by miquelbeltran** [comment]
I still fail to understand why do you want to mock the plugin implementation, instead of wrapping it and mocking your interface.

**Comment by ju851zel** [comment]
I'm using the Plugin in a service dedicated for getting device information, which already is the abstraction layer.
But when testing that service and the communication with the plugin mocking the `TargetPlatform` would be necessary.

The service for example has these two methods:

```dart
void doSomething() async {
  if (defaultTargetPlatform == TargetPlatform.android && await _isAndroidSdkVersionAbove30()) {
    doSomethingElse();
  } else {
    doSomethingCompletelyElse();
  }
}

Future<bool> _isAndroidSdkVersionAbove30() async =>
    (await DeviceInfoPlugin().androidInfo).version.sdkInt > 30;
```

Now I want to test the method `doSomething`.
Therefor I'd like to test if on iOS Devices the method `doSomethingCompletelyElse()` was called.
For this, in the test, I'm setting the `debugDefaultTargetPlatformOverride = TargetPlatform.iOS;` which works fine.

Now when trying to test the Android behaviour I'm setting `debugDefaultTargetPlatformOverride = TargetPlatform.android;`.
I also mock the platform Channel for the method `getDeviceInfo` of the Method channel `dev.fluttercommunity.plus/device_info`.

and the device_info_plugin calls internally this method.
```dart
@override
  Future<BaseDeviceInfo> deviceInfo() async {
    return BaseDeviceInfo(
        (await channel.invokeMethod('getDeviceInfo')).cast<String, dynamic>());
  }
```
(class `MethodChannelDeviceInfo` from the file `method_channel_device_info.dart`)

This works fine when running the tests on MacOS because on MacOS my mocked away `getDeviceInfo` platform method is called and returns my mocked object.
But on linux machines the mocked `getDeviceInfo` call does not get called, because on linux the method `(await channel.invokeMethod('getDeviceInfo'))` does not return a map of `<String,dynamic>` but simply a `String` and therefore an exception occurs.


As I'm thinking about that the `deviceInfo` method should not do a hard cast when certain plattforms do not return a `<String,dynamic>`. So this would be one way to solve the exception at least.

But my general question would still be the same, could we replace the `Platform.isXXX` with the `defaultTargetPlatform == TargetPlatform.XXX`.
Because this would not change the existing functionality, but only add the option to override the `defaultTargetPlatform` in tests.

I hope it get's clear what I'm trying to archive, Thanks in advance for reading this far 😅

**Comment by miquelbeltran** [comment]
IMO, you should mock your `_isAndroidSdkVersionAbove30` method rather than attempting on mocking the internals of a 3rd party library.

The problem is that you don't "own" the device_info_plus plugin, and so you are subject to any changes we may do in the internals at any given time.

You could even mock the implicit interface of the `DeviceInfoPlugin` class, and return your own hardcoded `DeviceInfoAndroid` when calling to `androidInfo`. That would also be a posibility.

But overall, I think it is the wrong approach to attempt to mock the internals of a 3rd party library.

**Comment by ju851zel** [comment]
You are right, mocking the `_isAndroidSdkVersionAbove30` would be sufficient for the layers above.
But testing the method itself would not be possible in this case.

You are also correct about the fact, that I don't own the plugin and cannot rely on any internals.

Nevertheless changing the platform determination via the `defaultTargetPlatform` would not affect the current implementation in a negative way, but create the possibility to mock it correctly.

**Comment by ju851zel** [comment]
**TL;DR on others having the same problem.**
 
The Method channel of the plugin can be overridden for the method `getDeviceInfo` and the tests can be configured, that on linux machines they are skipped.
This is necessary because on linux the method channel is not called.

On MacOS (other systems where not tested by myself) the method channel is called and therefore the tests run successfully.
